File: examples/ios/Examples/QueryEngine.m
#import <XCTest/XCTest.h>
#import <Realm/Realm.h>

// :code-block-start: models
// Task.h
@interface QueryEngineExamplesObjc_Task : RLMObject
@property NSString *name;
@property bool isComplete;
@property NSString *assignee;
@property int priority;
@property int progressMinutes;
@end
RLM_ARRAY_TYPE(QueryEngineExamplesObjc_Task)
// Task.m
@implementation QueryEngineExamplesObjc_Task
@end

// Project.h
@interface QueryEngineExamplesObjc_Project : RLMObject
@property NSString *name;
@property RLMArray<QueryEngineExamplesObjc_Task> *tasks;
@end
// Project.m
@implementation QueryEngineExamplesObjc_Project
@end
// :code-block-end:

// Be sure to differentiate test cases between Swift/Obj-C, or else
// the test explorer will get confused.
@interface QueryEngineObjc : XCTestCase

@end

@implementation QueryEngineObjc

- (void)tearDown {
    RLMRealm *realm = [RLMRealm defaultRealm];
    [realm transactionWithBlock:^() {
        [realm deleteAllObjects];
    }];
}

- (void)testPredicates {
    // :code-block-start: predicates
    NSPredicate *predicate = [NSPredicate predicateWithFormat:@"progressMinutes > %@ AND name == %@", @1, @"Ali"];
    // :code-block-end:
    
    // :code-block-start: substitutions
    [NSPredicate predicateWithFormat:@"%K > %@ AND %K == %@", @"progressMinutes", @1, @"name", @"Ali"];
    // :code-block-end:
}

- (void)testFilters {
    // :code-block-start: setup
    RLMRealm *realm = [RLMRealm defaultRealm];
    [realm transactionWithBlock:^() {
        // Add projects and tasks here
    }];
    
    RLMResults *tasks = [QueryEngineExamplesObjc_Task allObjectsInRealm:realm];
    RLMResults *projects = [QueryEngineExamplesObjc_Project allObjectsInRealm:realm];
    // :code-block-end:
    
    // :code-block-start: comparison-operators
    NSLog(@"High priority tasks: %lu",
          [[tasks objectsWithPredicate:[NSPredicate predicateWithFormat:@"priority > %@", @5]] count]);

    NSLog(@"Short running tasks: %lu",
          [[tasks objectsWhere:@"progressMinutes between {1, 15}"] count]);

    NSLog(@"Unassigned tasks: %lu",
          [[tasks objectsWhere:@"assignee == nil"] count]);

    NSLog(@"Ali or Jamie's tasks: %lu",
          [[tasks objectsWhere:@"assignee IN {'Ali', 'Jamie'}"] count]);
    
    NSLog(@"Tasks with progress between 30 and 60 minutes: %lu",
          [[tasks objectsWhere:@"progressMinutes BETWEEN {30, 60}"] count]);
    
    // :code-block-end:

    // :code-block-start: logical-operators
    NSLog(@"Ali's complete tasks: %lu",
      [[tasks objectsWhere:@"assignee == 'Ali' AND isComplete == true"] count]);
    // :code-block-end:
    
    // :code-block-start: string-operators
    // Use [c] for case-insensitivity.
    NSLog(@"Projects that start with 'e': %lu",
      [[projects objectsWhere:@"name BEGINSWITH[c] 'e'"] count]);

    NSLog(@"Projects that contain 'ie': %lu",
      [[projects objectsWhere:@"name CONTAINS 'ie'"] count]);
    // :code-block-end:
    
    // :code-block-start: aggregate-operators
    NSLog(@"Projects with average tasks priority above 5: %lu",
      [[projects objectsWhere:@"tasks.@avg.priority > 5"] count]);

    NSLog(@"Long running projects: %lu",
      [[projects objectsWhere:@"tasks.@sum.progressMinutes > 100"] count]);
    // :code-block-end:
    
    // :code-block-start: set-operators
    NSLog(@"Projects with no complete tasks: %lu",
      [[projects objectsWhere:@"NONE tasks.isComplete == true"] count]);

    NSLog(@"Projects with any top priority tasks: %lu",
      [[projects objectsWhere:@"ANY tasks.priority == 10"] count]);
    // :code-block-end:
    
    // :code-block-start: subquery
    NSPredicate *predicate = [NSPredicate predicateWithFormat:
                              @"SUBQUERY(tasks, $task, $task.isComplete == %@ AND $task.assignee == %@).@count > 0",
                              @NO,
                              @"Alex"];
    NSLog(@"Projects with incomplete tasks assigned to Alex: %lu",
      [[projects objectsWithPredicate:predicate] count]);
    // :code-block-end:
}

@end

File: examples/ios/Examples/QueryEngine.swift
// :replace-start: {
//   "terms": {
//     "QueryEngineExamples_": ""
//   }
// }
import XCTest
import RealmSwift

// :code-block-start: models
class QueryEngineExamples_Task: Object {
    @Persisted var name = ""
    @Persisted var isComplete = false
    @Persisted var assignee: String?
    @Persisted var priority = 0
    @Persisted var progressMinutes = 0
}

class QueryEngineExamples_Project: Object {
    @Persisted var name = ""
    @Persisted var tasks: List<QueryEngineExamples_Task>
}
// :code-block-end:

class QueryEngine: XCTestCase {
    override func tearDown() {
        let realm = try! Realm()
        try! realm.write {
            realm.deleteAll()
        }
    }

    func testPredicates() {
        // :code-block-start: predicates
        let predicate = NSPredicate(format: "progressMinutes > 1 AND name == %@", "Ali")
        // :code-block-end:

        // :code-block-start: substitutions
        NSPredicate(format: "%K > %@ AND %K == %@", "progressMinutes", NSNumber(1), "name", "Ali")
        // :code-block-end:
    }

    func testFilters() {
        // :code-block-start: setup
        let realm = try! Realm()
        try! realm.write {
            // Add tasks and projects here.
            let project = QueryEngineExamples_Project()
            project.name = "New Project"
            let task = QueryEngineExamples_Task()
            task.assignee = "Alex"
            task.priority = 5
            project.tasks.append(task)
            realm.add(project)
            // ...
        }
        let tasks = realm.objects(QueryEngineExamples_Task.self)
        let projects = realm.objects(QueryEngineExamples_Project.self)
        // :code-block-end:

        // :code-block-start: comparison-operators
        let highPriorityTasks = tasks.filter("priority > 5")
        print("High priority tasks: \(highPriorityTasks.count)")

        let longRunningTasks = tasks.filter("progressMinutes > 120")
        print("Long running tasks: \(longRunningTasks.count)")

        let unassignedTasks = tasks.filter("assignee == nil")
        print("Unassigned tasks: \(unassignedTasks.count)")

        let aliOrJamiesTasks = tasks.filter("assignee IN {'Ali', 'Jamie'}")
        print("Ali or Jamie's tasks: \(aliOrJamiesTasks.count)")

        let progressBetween30and60 = tasks.filter("progressMinutes BETWEEN {30, 60}")
        print("Tasks with progress between 30 and 60 minutes: \(progressBetween30and60.count)")
        // :code-block-end:

        // :code-block-start: logical-operators
        let aliComplete = tasks.filter("assignee == 'Ali' AND isComplete == true")
        print("Ali's complete tasks: \(aliComplete.count)")
        // :code-block-end:

        // :code-block-start: string-operators
        // Use [c] for case-insensitivity.
        let startWithE = projects.filter("name BEGINSWITH[c] 'e'")
        print("Projects that start with 'e': \(startWithE.count)")

        let containIe = projects.filter("name CONTAINS 'ie'")
        print("Projects that contain 'ie': \(containIe.count)")

        // [d] for diacritic insensitivty: contains 'e', 'E', 'é', etc.
        let containElike = projects.filter("name CONTAINS[cd] 'e'")
        print("Projects that contain 'e', 'E', 'é', etc.: \(containElike.count)")
        // :code-block-end:

        // :code-block-start: aggregate-operators
        let averageTaskPriorityAbove5 = projects.filter("tasks.@avg.priority > 5")
        print("Projects with average task priority above 5: \(averageTaskPriorityAbove5.count)")

        let allTasksLowerPriority = projects.filter("tasks.@max.priority < 5")
        print("Projects where all tasks are lower priority: \(allTasksLowerPriority.count)")

        let allTasksHighPriority = projects.filter("tasks.@min.priority > 5")
        print("Projects where all tasks are high priority: \(allTasksHighPriority.count)")

        let moreThan5Tasks = projects.filter("tasks.@count > 5")
        print("Projects with more than 5 tasks: \(moreThan5Tasks.count)")

        let longRunningProjects = projects.filter("tasks.@sum.progressMinutes > 100")
        print("Long running projects: \(longRunningProjects.count)")
        // :code-block-end:

        // :code-block-start: set-operators
        let noCompleteTasks = projects.filter("NONE tasks.isComplete == true")
        print("Projects with no complete tasks: \(noCompleteTasks.count)")

        let anyTopPriorityTasks = projects.filter("ANY tasks.priority == 10")
        print("Projects with any top priority tasks: \(anyTopPriorityTasks.count)")
        // :code-block-end:

        // :code-block-start: subquery
        let predicate = NSPredicate(
            format: "SUBQUERY(tasks, $task, $task.isComplete == false AND $task.assignee == %@).@count > 0", "Alex")
        print("Projects with incomplete tasks assigned to Alex: \(projects.filter(predicate).count)")
        // :code-block-end:
    }
}

// :replace-end:
